### PR TITLE
feat: add packed mersenne31

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR adds PackedPrimeField for `Mersenne31`.

### CODE REVIEW

File: tachyon/base/strings/string_util.h [hunk 1]
```diff
@@ -146,11 +146,11 @@ std::string VectorToString(const std::vector<T>& data) {
 }
 
 template <typename T>
-std::string Vector2DToString(const std::vector<std::vector<T>>& data) {
+std::string Container2DToString(const std::vector<std::vector<T>>& data) {
```
Comment (ashjeong) [inline tachyon/base/strings/string_util.h:149]: Should `T` be changed to `container` and this function intake a container of `containers`?

File: tachyon/math/finite_fields/mersenne31/packed_mersenne31_neon.cc [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+  //      mul      t.4s, lhs.4s, rhs.4s
+  //      mla      t.4s, prod_hi31.4s, P.4s
+  //      sub      u.4s, t.4s, P.4s
+  //      umin     r.4s, t.4s, u.4s
+  // throughput: 1.25 cyc/vec (3.2 els/cyc)
+  // latency: 10 cyc
+
+  // Let prod := lhs * rhs
+  //     prod_hi31 := (prod >> 2³¹) (mod 2³¹)
+  //               = a₃₁ + 2 * a₃₂ + 2² * a₃₃ + ... + 2³⁰ * a₆₂
```
Comment (Insun35) [inline tachyon/math/finite_fields/mersenne31/packed_mersenne31_neon.cc:60]: I believe the superscript unicode of zero in 2^30 went wrong. It appears just like a tiny box for me. Please check!

**Review by Insun35 — COMMENTED** [review]
Please fix the typo in de14c3b commit body:
Aditionally -> Additionally

File: tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+#include <immintrin.h>
+
+#include "tachyon/math/finite_fields/packed_prime_field32_avx2.h"
+
+namespace tachyon::math {
+
+namespace {
+
+__m256i kP;
```
Comment (Insun35) [inline tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc:16]: (Optional) what about `kModulus` for clarity?

File: tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx512.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+  constexpr static size_t N = 16;
+};
+
+class PackedMersenne31AVX512
+    : public PackedPrimeFieldBase<PackedMersenne31AVX512> {
+ public:
+  using PrimeField = Mersenne31;
+
+  constexpr static size_t N = 16;
```
Comment (dongchangYoo) [inline tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx512.h:30]: How about setting `N` from PackedPrimeFieldTraits.
It seems that `N` has been defined multiple times

File: tachyon/math/finite_fields/mersenne31/packed_mersenne31_neon.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  using PrimeField = Mersenne31;
+
+  constexpr static size_t N = 4;
+};
+
+class PackedMersenne31Neon : public PackedPrimeFieldBase<PackedMersenne31Neon> {
+ public:
+  using PrimeField = Mersenne31;
+
+  constexpr static size_t N = 4;
```
Comment (dongchangYoo) [inline tachyon/math/finite_fields/mersenne31/packed_mersenne31_neon.h:29]: It seems that `N` has been defined multiple times.

**Review by dongchangYoo — CHANGES_REQUESTED** [review]

File: tachyon/math/finite_fields/mersenne31/BUILD.bazel [hunk 6]
```diff
@@ -12,3 +26,51 @@ generate_prime_fields(
...
+    hdrs = ["packed_mersenne31.h"],
+    defines = tachyon_avx512_defines(),
+    deps = [
+        ":packed_mersenne31_neon",
+        "//tachyon/build:build_config",
+    ] + if_has_avx512(
+        [":packed_mersenne31_avx512"],
+        [":packed_mersenne31_avx2"],
+    ),
+)
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/math/finite_fields/mersenne31/BUILD.bazel:76]: Oh this is intended, please refer to below!

```bzl
def if_has_avx512(a, b = []):
    return select({
        "@kroma_network_tachyon//:tachyon_has_avx512": a,
        "//conditions:default": b,
    })
```

File: tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc (hunk 3 above)
Comment (chokobole) [reply tachyon/math/finite_fields/mersenne31/packed_mersenne31_avx2.cc:16]: the reason why I chose `P` are 2.

1. To match the original code as much as possible
2. The variable in the comment is a just single letter.

**Review by Insun35 — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**